osu.Game/Screens/Select/FilterControl.cs

```diff
@@ -64,7 +65,7 @@ public FilterCriteria CreateCriteria()
                 Sort = sortMode.Value,
                 AllowConvertedBeatmaps = showConverted.Value,
                 Ruleset = ruleset.Value,
-                CollectionBeatmapMD5Hashes = collectionDropdown.Current.Value?.Collection?.PerformRead(c => c.BeatmapMD5Hashes)
+                CollectionBeatmapMD5Hashes = collectionDropdown.Current.Value?.Collection?.PerformRead(c => c.BeatmapMD5Hashes).ToList()
```

Comment:
At the point this is being forcibly materialised, might as well do it to a type that is designed for this sort of usage? By which I mean

```suggestion
                CollectionBeatmapMD5Hashes = collectionDropdown.Current.Value?.Collection?.PerformRead(c => c.BeatmapMD5Hashes).ToHashSet()
```

or even

```suggestion
                CollectionBeatmapMD5Hashes = collectionDropdown.Current.Value?.Collection?.PerformRead(c => c.BeatmapMD5Hashes).ToImmutableHashSet()
```

I'm not sure if things are going to be smart enough to work with just that given that `CollectionBeatmapMD5Hashes` is just an `IEnumerable<string>`, but maybe? (and the type could always be changed if using hashset has no gains when calling `.Contains()` through `IEnumerable`)

Reply:
In fact, after testing with the database provided [here](https://github.com/ppy/osu/issues/27322#issuecomment-1960943814), as far as I can tell, using `.ToList()` actually makes the performance of this _much worse_ in extreme cases unless `ImmutableHashSet` is used!

| master | this PR | this PR but `ImmutableHashSet` |
| :-: | :-: | :-: |
| ![master](https://github.com/ppy/osu/assets/20418176/b0dcb2c2-952b-4646-a80e-8a856af0ec6e) | ![list](https://github.com/ppy/osu/assets/20418176/7a26c50f-b9d9-4646-a579-a86c63856dd5) | ![hashset](https://github.com/ppy/osu/assets/20418176/21661633-1357-4917-a9ba-53ed3cf0cfd8) |

Note that the list variant takes _literal minutes_ to complete.

Reply:
Ah yep, that's an oversight. Was waiting to have data to test with to confirm results.